### DISCUSSION TITLE
Unable to delete Zone

### DISCUSSION THREAD

**Discussion by mimcdevitt** [apache/cloudstack#9532] [Q&A] [answered]
Hello, I tried to delete one of our zones and have gotten stuck.   I deleted all hosts, pods, guest and physical networks but forgot to delete templates first.   Then I was unable to remove zone due to a couple hundred templates in zone secondary storage,  I am unable to delete templates as secondary storage vm is no longer running and there is no network associated with this zone.  Is their a way to recover from this and get the zone and templates removed.    Thank you

**Comment by Pearl1594** [comment, marked as answer]
If you are trying to decommission the zone, and this is a test environment, then you could mark all the templates of the zone as Inactive in the database. To get the list of all the Active templates 
```
select distinct(id) from template_view where store_id in (select id from image_store where data_center_id = <id_of_zone>) and type = 'USER' and template_state = 'Active'\G
```

And mark their states as Inactive
```
update vm_template set state = 'Inactive' where id in (select distinct(id) from template_view where store_id in (select id from image_store where data_center_id = <id_of_zone>) and type = 'USER' and template_state = 'Active'\G);
```
You can then manually clean up your secondary store. 

Please take note, that this is more of a hack to clean up your env - do it only if it's a test environment.

**Reply by mimcdevitt** [reply]
Thank you very much,  this is a test environment.  Will give that suggestion a try.

**Reply by jack99trade** [reply]
Hi @Pearl1594 I am also getting error with POD deletion :-  " Request failed. (530) The pod cannot be deleted because there are servers in this pod.  "

I have deleted  All the instances, ISOs , templates , KVM host , Cluster also deleted physical interface 1 of Zone.   but now error in POD. 
my objective is to get rid of this test zone I created.

**Reply by Pearl1594** [reply]
@jack99trade Based on the error message, it indicates that the pod has hosts - these hosts could be physical hosts or even SSVM / CPVM. You can check it be running the following:
```
select * from host where pod_id = <ID_of_pod> where removed is null \G
```
If you see any un-removed host, mark the removed field with the current time

```
update host set removed = now() where id = (select id from host where pod_id = <ID_of_pod> where removed is null);
```

**Reply by mimcdevitt** [reply]
I can get the list of templates but when I try to update to inactive I get the following error.   
SQL Error 1443: The definition of table template_view prevents operation UPDATE on table 'vm_template'

Should the update be run on a another table rather than that one

**Reply by mimcdevitt** [reply]
Hi @Pearl1594 do you have any suggestions for my issue of being unable to update status of Templates to Inactive
Thanks
Mike

**Reply by jack99trade** [reply]
> @jack99trade Based on the error message, it indicates that the pod has hosts - these hosts could be physical hosts or even SSVM / CPVM. You can check it be running the following:
> 
> ```
> select * from host where pod_id = <ID_of_pod> where removed is null \G
> ```
> 
> If you see any un-removed host, mark the removed field with the current time
> 
> ```
> update host set removed = now() where id = (select id from host where pod_id = <ID_of_pod> where removed is null);
> ```

Thank you @Pearl1594  I have done this and finally its gone :)

**Reply by jack99trade** [reply]
> I can get the list of templates but when I try to update to inactive I get the following error. SQL Error 1443: The definition of table template_view prevents operation UPDATE on table 'vm_template'
> 
> Should the update be run on a specific table rather than a view

Hi @mimcdevitt  i think you can try with first part  :  update vm_template set state = 'Inactive' where id   :   for one of the IDs you found from step1.   I also did the same for my POD deletion.

**Reply by mimcdevitt** [reply]
Thanks,  I can try with a single record but I have hundreds to do;  I think I am running into an issue  where MySQL doesn't allow updates to a table when the update involves a subquery that selects from a view that references the same table. This creates a situation where the update operation could potentially create inconsistent or unexpected results.   So I am going to try storing the results of the select in a temp table and run the update from the data in that table.

**Reply by jack99trade** [reply]
> Thanks, I can try with a single record but I have hundreds to do; I think I am running into an issue where MySQL doesn't allow updates to a table when the update involves a subquery that selects from a view that references the same table. This creates a situation where the update operation could potentially create inconsistent or unexpected results. So I am going to try storing the results of the select in a temp table and run the update from the data in that table.

yeah that sounds like a good plan and I agree with you :)

**Reply by rajujith** [reply]
@mimcdevitt  I am assuming that you have more than one zone, then you might want to skip updating the vm_template for the cross-zone templates.
